File: android/src/main/kotlin/dev/aws/ivs_realtime/AwsIvsRealtimePlugin.kt
package dev.aws.ivs_realtime

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry

class AwsIvsRealtimePlugin : FlutterPlugin, ActivityAware {

    private var flutterPluginBinding: FlutterPlugin.FlutterPluginBinding? = null
    private var methodChannel: MethodChannel? = null
    private var stageEventChannel: EventChannel? = null
    private var stageEventSink: EventChannel.EventSink? = null
    private var controller: IvsStageController? = null
    private var activityBinding: ActivityPluginBinding? = null
    private var permissionsListener: PluginRegistry.RequestPermissionsResultListener? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        flutterPluginBinding = binding
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE,
            IvsStageViewFactory { controller ?: error("Activity not attached yet") },
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel?.setMethodCallHandler(null)
        methodChannel = null
        stageEventChannel?.setStreamHandler(null)
        stageEventChannel = null
        stageEventSink = null
        flutterPluginBinding = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
        val act = binding.activity
        val messenger = flutterPluginBinding!!.binaryMessenger

        stageEventChannel =
            EventChannel(messenger, EVENTS_CHANNEL).also { ec ->
                ec.setStreamHandler(
                    object : EventChannel.StreamHandler {
                        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                            stageEventSink = events
                        }

                        override fun onCancel(arguments: Any?) {
                            stageEventSink = null
                        }
                    },
                )
            }

        val c =
            IvsStageController(act) { map ->
                stageEventSink?.success(map)
            }
        controller = c
        val listener =
            PluginRegistry.RequestPermissionsResultListener { requestCode, permissions, grantResults ->
                c.onRequestPermissionsResult(requestCode, permissions, grantResults)
            }
        permissionsListener = listener
        binding.addRequestPermissionsResultListener(listener)
        methodChannel =
            MethodChannel(messenger, CHANNEL).also { ch ->
                ch.setMethodCallHandler { call, result ->
                    when (call.method) {
                        "join" -> {
                            val args = call.arguments as? Map<*, *>
                            val token = args?.get("token") as? String ?: ""
                            val publish = channelBoolean(args?.get("publish"), default = true)
                            c.setPublishEnabled(publish)
                            c.joinOrLeave(token, result)
                        }
                        "leave" -> {
                            c.leave()
                            result.success(null)
                        }
                        "setPublish" -> {
                            val enabled = channelBoolean(call.arguments, default = true)
                            c.setPublishEnabled(enabled)
                            result.success(null)
                        }
                        "refreshStageBindings" -> {
                            c.refreshStageBindings()
                            result.success(null)
                        }
                        "setLocalStreamMuted" -> {
                            val m = call.arguments as? Map<*, *>
                            val mic = m?.get("micMuted") as? Boolean ?: false
                            val cam = m?.get("cameraMuted") as? Boolean ?: false
                            c.setLocalStreamMuted(micMuted = mic, cameraMuted = cam)
                            result.success(null)
                        }
                        "setShowParticipantStateOverlay" -> {
                            val m = call.arguments as? Map<*, *>
                            val visible = channelBoolean(m?.get("visible"), default = false)
                            c.setShowParticipantStateOverlay(visible)
                            result.success(null)
                        }
                        else -> result.notImplemented()
                    }
                }
            }
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        val binding = activityBinding
        val listener = permissionsListener
        if (binding != null && listener != null) {
            binding.removeRequestPermissionsResultListener(listener)
        }
        activityBinding = null
        permissionsListener = null
        controller?.release()
        controller = null
        methodChannel?.setMethodCallHandler(null)
        methodChannel = null
        stageEventChannel?.setStreamHandler(null)
        stageEventChannel = null
        stageEventSink = null
    }

    private fun channelBoolean(value: Any?, default: Boolean): Boolean =
        when (value) {
            is Boolean -> value
            is Number -> value.toInt() != 0
            null -> default
            else -> default
        }

    companion object {
        const val CHANNEL = "aws_ivs_realtime/stage"
        const val EVENTS_CHANNEL = "aws_ivs_realtime/stage_events"
        const val VIEW_TYPE = "ivs_stage_view"
    }
}

File: ios/Classes/AwsIvsRealtimePlugin.swift
import Flutter
import UIKit

private enum AwsIvsRealtimeChannels {
  static let stage = "aws_ivs_realtime/stage"
  static let stageEvents = "aws_ivs_realtime/stage_events"
  static let viewType = "ivs_stage_view"
}

private final class StageEventsStreamHandler: NSObject, FlutterStreamHandler {
  weak var controller: IvsStageController?

  init(controller: IvsStageController) {
    self.controller = controller
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    controller?.eventSink = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    controller?.eventSink = nil
    return nil
  }
}

public class AwsIvsRealtimePlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let controller = IvsStageController()

    let eventChannel = FlutterEventChannel(
      name: AwsIvsRealtimeChannels.stageEvents,
      binaryMessenger: registrar.messenger()
    )
    eventChannel.setStreamHandler(StageEventsStreamHandler(controller: controller))

    let channel = FlutterMethodChannel(
      name: AwsIvsRealtimeChannels.stage,
      binaryMessenger: registrar.messenger()
    )
    registrar.register(
      IvsStageViewFactory(controller: controller),
      withId: AwsIvsRealtimeChannels.viewType
    )
    channel.setMethodCallHandler { call, result in
      switch call.method {
      case "join":
        let args = call.arguments as? [String: Any]
        let token = args?["token"] as? String ?? ""
        let publish = Self.channelBool(args?["publish"], defaultVal: true)
        controller.setPublishEnabled(publish)
        controller.joinOrLeave(token: token, result: result)
      case "leave":
        controller.leave()
        result(nil)
      case "setPublish":
        let enabled = Self.channelBool(call.arguments, defaultVal: true)
        controller.setPublishEnabled(enabled)
        result(nil)
      case "refreshStageBindings":
        controller.refreshStageBindings()
        result(nil)
      case "setLocalStreamMuted":
        let m = call.arguments as? [String: Any]
        let mic = m?["micMuted"] as? Bool ?? false
        let cam = m?["cameraMuted"] as? Bool ?? false
        controller.setLocalStreamMuted(micMuted: mic, cameraMuted: cam)
        result(nil)
      case "setShowParticipantStateOverlay":
        let m = call.arguments as? [String: Any]
        let visible = Self.channelBool(m?["visible"], defaultVal: false)
        controller.setShowParticipantStateOverlay(visible)
        result(nil)
      default:
        result(FlutterMethodNotImplemented)
      }
    }

    NotificationCenter.default.addObserver(
      forName: UIApplication.willTerminateNotification,
      object: nil,
      queue: .main
    ) { _ in
      controller.releaseStage()
    }
  }

  private static func channelBool(_ value: Any?, defaultVal: Bool) -> Bool {
    switch value {
    case let b as Bool:
      return b
    case let n as NSNumber:
      return n.intValue != 0
    case nil:
      return defaultVal
    default:
      return defaultVal
    }
  }
}
